codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -66,412 +70,1049 @@
+    }
+
+    Http2MultiplexCodec(Http2ConnectionEncoder encoder, Http2ConnectionDecoder decoder, Http2Settings initialSettings,
+                    long gracefulShutdownTimeoutMillis,  ChannelHandler inboundStreamHandler) {
+        super(encoder, decoder, initialSettings, gracefulShutdownTimeoutMillis);
+        this.inboundStreamHandler = inboundStreamHandler;
+    }
+
+    static ChannelHandler checkSharable(ChannelHandler handler) {
+        if ((handler instanceof ChannelHandlerAdapter && !((ChannelHandlerAdapter) handler).isSharable()) ||
+                !handler.getClass().isAnnotationPresent(Sharable.class)) {
+            throw new IllegalArgumentException("The handler must be Sharable");
+        }
+        return handler;
+    }
+
+    private static void registerDone(ChannelFuture future) {
+        // Handle any errors that occurred on the local thread while registering. Even though
+        // failures can happen after this point, they will be handled by the channel by closing the
+        // childChannel.
+        DefaultHttp2StreamChannel childChannel = (DefaultHttp2StreamChannel) future.channel();
+        if (!future.isSuccess()) {
+            if (childChannel.isRegistered()) {
+                childChannel.close();
+            } else {
+                childChannel.unsafe().closeForcibly();
+            }
         }
-        this.server = server;
-        this.bootstrap = new Http2StreamChannelBootstrap(bootstrap);
     }
 
     @Override
-    public void handlerAdded(ChannelHandlerContext ctx) {
+    public final void handlerAdded0(ChannelHandlerContext ctx) throws Exception {
+        if (ctx.executor() != ctx.channel().eventLoop()) {
+            throw new IllegalStateException("EventExecutor must be EventLoop of Channel");
+        }
         this.ctx = ctx;
-        bootstrap.parentChannel(ctx.channel());
     }
 
     @Override
-    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
-        if (!(cause instanceof StreamException)) {
-            ctx.fireExceptionCaught(cause);
-            return;
-        }
+    public final void handlerRemoved0(ChannelHandlerContext ctx) throws Exception {
+        super.handlerRemoved0(ctx);
 
-        StreamException streamEx = (StreamException) cause;
-        try {
-            Http2StreamChannel childChannel = childChannels.get(streamEx.streamId());
-            if (childChannel != null) {
-                childChannel.pipeline().fireExceptionCaught(streamEx);
-            } else {
-                logger.warn(format("Exception caught for unknown HTTP/2 stream '%d'", streamEx.streamId()),
-                            streamEx);
-            }
-        } finally {
-            onStreamClosed(streamEx.streamId());
+        DefaultHttp2StreamChannel ch = head;
+        while (ch != null) {
+            DefaultHttp2StreamChannel curr = ch;
+            ch = curr.next;
+            curr.next = null;
```

Comment:
I guess the iteration is done in attempt to avoid GC nepotism? Consider adding a comment if this is the case.